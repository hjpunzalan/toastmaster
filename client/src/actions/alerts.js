import { SET_ALERT, RESET_ALERT } from './types';

export const setAlert = (msg, alertType) => dispatch => {
	dispatch({
		type: SET_ALERT,
		payload: { msg, alertType }
	});
};

export const resetAlert = () => dispatch => {
	dispatch({ type: RESET_ALERT });
};
