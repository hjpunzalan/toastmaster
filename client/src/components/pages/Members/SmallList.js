import React, { useEffect } from 'react';
import { MdEmail } from 'react-icons/md';
import img from '../../../img/anonymous.png';

const SmallList = ({
	users,
	Moderator,
	active,
	deActivateUser,
	activateUser,
	currentUser,
	setActive,
	setAlert,
	resetAlert,
	changeRole
}) => {
	useEffect(() => {
		if (users.length === 0) {
			resetAlert();
			setActive(true);
			setAlert('Currently there are no unactive users', 'fail');
		}
		// eslint-disable-next-line
	}, [users]);

	return (
		<div className="SmallList">
			{users.map(user => (
				<div className="SmallList__user" key={user._id}>
					<img
						className="SmallList__photo"
						src={user.photo || img}
						alt="avatar"
					/>
					<div className="SmallList__body">
						<div className="SmallList__details">
							<span className="SmallList__name">
								{user.firstName}&nbsp;{user.lastName}
							</span>
							{(active || Moderator) && (
								<a
									className="SmallList__email"
									href={`mailto:${user.email}`}
									target="_top">
									<MdEmail />
									<span>{user.email}</span>
								</a>
							)}
						</div>
						{Moderator && (
							<div className="SmallList__moderator">
								{Moderator && currentUser.role === 'admin' && active && (
									<div className="SmallList__committee">
										<strong>Committee</strong>
										<select
											defaultValue={
												user.role === 'committee' || user.role === 'admin'
											}
											onChange={e => changeRole(user._id, e.target.value)}>
											<option value={true}>Yes</option>
											<option value={false}>No</option>
										</select>
									</div>
								)}
								{active ? (
									<div
										className="SmallList__deactivate"
										onClick={() => deActivateUser(user._id)}>
										Deactivate
									</div>
								) : (
									<div
										className="SmallList__activate"
										onClick={() => activateUser(user._id)}>
										Activate
									</div>
								)}
							</div>
						)}
					</div>
				</div>
			))}
		</div>
	);
};

export default SmallList;
