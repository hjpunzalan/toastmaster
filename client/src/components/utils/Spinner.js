import React, { Fragment } from 'react';
import spinner from '../../img/spinner.gif';

export default () => (
	<Fragment>
		<img
			src={spinner}
			style={{ width: '150px', margin: '20% auto 0 auto', display: 'block' }}
			alt="Loading..."
		/>
	</Fragment>
);
