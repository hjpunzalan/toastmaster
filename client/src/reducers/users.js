import {
	REGISTER_SUCCESS,
	GET_ALL_USERS,
	TOGGLE_MODERATOR,
	USER_ERROR,
	DEACTIVATE_USER,
	ACTIVATE_USER,
	LOADING_USER
} from '../actions/types';

const initialState = {
	users: [],
	loading: true,
	error: {},
	Moderator: false
};

export default (state = initialState, action) => {
	const { type, payload } = action;

	const replaceUser = (id, userArray, newUser) => {
		const index = userArray.findIndex(el => el._id === id);
		userArray[index] = newUser;
		return userArray;
	};

	switch (type) {
		case USER_ERROR:
			return { ...state, loading: false };
		case LOADING_USER:
			return { ...state, loading: true };
		case REGISTER_SUCCESS:
			localStorage.setItem(
				'users',
				JSON.stringify([...state.users, ...payload])
			);
			return {
				//State needs to be first
				...state,
				users: [...state.users, ...payload],
				loading: false
			};
		case GET_ALL_USERS:
			return {
				...state,
				loading: false,
				users: payload
			};
		case TOGGLE_MODERATOR:
			return {
				...state,
				Moderator: !state.Moderator
			};

		case DEACTIVATE_USER:
		case ACTIVATE_USER:
			return {
				...state,
				users: replaceUser(payload._id, state.users, payload),
				loading: false
			};

		default:
			return state;
	}
};
