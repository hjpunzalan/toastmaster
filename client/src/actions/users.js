import { REGISTER_SUCCESS } from '../actions/types';
import { setAlert, resetAlert } from './alerts';

export const registerUser = formData => dispatch => {
	const { firstName, lastName, email } = formData;
	dispatch(resetAlert());
	const users = [
		{
			firstName: 'Jonathan',
			lastName: 'Punzalan',
			email: 'hj.punzalan@hotmail.com'
		}
	];

	const checkUsers = users.filter(user => user.email === email);

	if (checkUsers.length === 0) {
		dispatch({
			type: REGISTER_SUCCESS,
			payload: formData
		});

		dispatch(
			setAlert(`${firstName} ${lastName} succesfully registered.`, 'success')
		);
	} else
		dispatch(setAlert(`Email: ${email} has already been registered`, 'fail'));
};
