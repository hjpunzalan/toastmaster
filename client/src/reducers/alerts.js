import { SET_ALERT, RESET_ALERT, ERROR } from "../actions/types";

export const initialState = {
	msg: [],
	alertType: null,
	error: {},
};

const alertsReducer = (state = initialState, action) => {
	const { type, payload } = action;
	switch (type) {
		case SET_ALERT:
			return {
				...state,
				msg: [...state.msg, payload.msg],
				alertType: payload.alertType,
			};
		case RESET_ALERT:
			return {
				...initialState,
			};
		case ERROR:
			return {
				...state,
				error: payload,
			};

		default:
			return state;
	}
};

export default alertsReducer;
