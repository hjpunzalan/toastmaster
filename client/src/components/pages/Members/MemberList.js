import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import {
	getAllUsers,
	deActivateUser,
	activateUser
} from '../../../actions/users';
import Table from './Table';
import Spinner from '../../utils/Spinner';

const MemberList = ({
	getAllUsers,
	users: { users, loading, Moderator },
	deActivateUser,
	activateUser
}) => {
	useEffect(() => {
		getAllUsers(active);
		//eslint-disable-next-line
	}, []);

	const [active, setActive] = useState(true);
	const activeUsers = users.filter(user => user.active === true);
	const unActiveUsers = users.filter(user => user.active === false);

	return loading ? (
		<Spinner />
	) : (
		<div className="MemberList">
			<h1 className="MemberList__title">Member List</h1>
			<div className="MemberList__top">
				<Link className="MemberList__register" to="/register">
					{Moderator && <button>Register a User</button>}
				</Link>

				<div className="MemberList__switch">
					<input
						type="radio"
						name="users"
						id="active"
						onChange={() => setActive(!active)}
						checked={active}
					/>
					<label htmlFor="active">Active</label>
					<input
						type="radio"
						name="users"
						id="unactive"
						onChange={() => setActive(!active)}
						checked={!active}
					/>
					<label htmlFor="unactive">Unactive</label>
				</div>
			</div>
			<Table
				Moderator={Moderator}
				users={active ? activeUsers : unActiveUsers}
				active={active}
				deActivateUser={deActivateUser}
				activateUser={activateUser}
			/>
		</div>
	);
};

MemberList.propTypes = {
	getAllUsers: PropTypes.func.isRequired,
	users: PropTypes.object.isRequired,
	deActivateUser: PropTypes.func.isRequired,
	activateUser: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
	users: state.users
});

export default connect(
	mapStateToProps,
	{ getAllUsers, deActivateUser, activateUser }
)(MemberList);
