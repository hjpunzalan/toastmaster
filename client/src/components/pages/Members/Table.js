import React from 'react';
import Members from './Members';

const Table = ({ users, Moderator, active, deActivateUser, activateUser }) => {
	return (
		<table className="MemberList__list">
			<thead>
				<tr>
					<th>Photo</th>
					<th>First Name</th>
					<th>Last Name</th>
					{(Moderator || active) && <th>Email</th>}
					{Moderator && (active ? <th>Deactivate</th> : <th>Activate</th>)}
				</tr>
			</thead>
			<tbody>
				{users.map(user => (
					<Members
						key={user._id}
						user={user}
						Moderator={Moderator}
						active={active}
						deActivateUser={deActivateUser}
						activateUser={activateUser}
					/>
				))}
			</tbody>
		</table>
	);
};

export default Table;
