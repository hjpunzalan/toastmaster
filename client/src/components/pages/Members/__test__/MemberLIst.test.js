
import React from "react";
import { BrowserRouter } from "react-router-dom";
import renderer from 'react-test-renderer';
import { shallow } from "enzyme";
import { createBrowserHistory } from "history";
import { getAllUsers,deActivateUser,activateUser,changeRole } from "../../../../actions/users";
import { setAlert,resetAlert } from "../../../../actions/alerts";
import {MemberList} from "../MemberList";
import Spinner from "../../../utils/Spinner";
import Table from '../Table'
import SmallList from '../SmallList'


const history = createBrowserHistory();
jest.spyOn(history, "push");

const initialProps = {
    users:{ users:[{
                _id: "test",
                 firstName: "test",
            }
        ],
        loading: false,
        Moderator: false
    },
    auth: {
        currentUser: {
            active: true,
            role: "user",
            _id: "test",
            firstName: "test",
            lastName: "test",
           email:"user@test.com"
            
    }},
    getAllUsers,
    deActivateUser,
    activateUser,
    setAlert,
    resetAlert,
    changeRole
}

const setup = (props = {}) => {
	const setupProps = { ...initialProps, ...props };
	return shallow(<MemberList {...setupProps} />);
};

test('render MemberList component', () => {
    const wrapper = setup();
    expect(wrapper.find(".MemberList").length).toBe(1)
})

test("Snapshot renders correctly", () => {
	const tree = renderer
		.create(<BrowserRouter><MemberList {...initialProps} /></BrowserRouter>)
		.toJSON();
	expect(tree).toMatchSnapshot()

})

test('should render spinner if loading is true', () => {
    const wrapper = setup({
        users: {
            users: [{
                _id: "test",
                 firstName: "test",
            }
        ],
        loading: true,
        Moderator: false
    }});
    expect(wrapper.find(".MemberList").length).toBe(0)
    expect(wrapper.contains(<Spinner />)).toBe(true)
})

test('should render Table and SmallList component', () => {
    const wrapper = setup();
     
    expect(wrapper.containsMatchingElement(<Table />)).toBe(true)
    expect(wrapper.containsMatchingElement(<SmallList />)).toBe(true)
})
