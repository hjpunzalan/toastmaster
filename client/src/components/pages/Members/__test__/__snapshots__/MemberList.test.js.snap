// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot renders correctly 1`] = `
<div
  className="MemberList"
>
  <h1
    className="MemberList__title"
  >
    Member List
  </h1>
  <div
    className="MemberList__top"
  >
    <a
      className="MemberList__register"
      href="/register"
      onClick={[Function]}
    />
    <div
      className="MemberList__switch"
    >
      <input
        checked={true}
        id="active"
        name="users"
        onChange={[Function]}
        type="radio"
      />
      <label
        htmlFor="active"
      >
        Active
      </label>
      <input
        checked={false}
        id="unactive"
        name="users"
        onChange={[Function]}
        type="radio"
      />
      <label
        htmlFor="unactive"
      >
        Unactive
      </label>
    </div>
  </div>
  <table
    className="MemberList__list"
  >
    <thead>
      <tr>
        <th>
          Photo
        </th>
        <th>
          First Name
        </th>
        <th>
          Last Name
        </th>
        <th>
          Email
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <div
    className="SmallList"
  />
</div>
`;
